gadgets/trace_ssl/program.bpf.c

```diff
@@ -0,0 +1,406 @@
+	if (tsp == 0)
+		return 0;
+	bpf_map_delete_elem(&ssl_starttime, &tid);
+
+	buf_copy_size = (size_t)MAX_BUF_SIZE < (size_t)len ?
+				(size_t)MAX_BUF_SIZE :
+				(size_t)len;
+	event = gadget_reserve_buf(&events, sizeof(struct event));
+	if (!event)
+		return 0;
+
+	event->mntns_id = mntns_id;
+	event->operation = op;
+	event->timestamp = ts;
+	event->latency_ns = ts - *tsp;
+	event->pid = pid;
+	event->tid = tid;
+	event->uid = (u32)bpf_get_current_uid_gid();
+	event->len = (u32)len;
+	event->retval = PT_REGS_RC(ctx);
+
+	bpf_get_current_comm(&event->comm, sizeof(event->comm));
+
+	if (!record_data ||
+	    bpf_probe_read_user(&event->buf, buf_copy_size, (char *)*bufp))
+		buf_copy_size = 0;
+
+	gadget_submit_buf(ctx, &events, event, EVENT_SIZE(buf_copy_size));
+
+	return 0;
+}
+
+#define PROBE_SSL_RW_ENTER(lib, func)                                       \
+	SEC("uprobe/" #lib ":" #func)                                       \
+	int trace_uprobe_##lib##_##func(struct pt_regs *ctx)                \
+	{                                                                   \
+		return probe_ssl_rw_enter(ctx, (void *)PT_REGS_PARM2(ctx)); \
+	}
+
+#define PROBE_SSL_RW_EXIT(lib, func)                            \
+	SEC("uretprobe/" #lib ":" #func)                        \
+	int trace_uretprobe_##lib##_##func(struct pt_regs *ctx) \
+	{                                                       \
+		return probe_ssl_rw_exit(ctx, lib##_##func);    \
+	}
+
+/* uprobes for libssl */
+PROBE_SSL_RW_ENTER(libssl, SSL_write)
+PROBE_SSL_RW_EXIT(libssl, SSL_write)
+
+PROBE_SSL_RW_ENTER(libssl, SSL_read)
+PROBE_SSL_RW_EXIT(libssl, SSL_read)
+
+SEC("uprobe/libssl:SSL_do_handshake")
+int trace_uprobe_libssl_SSL_do_handshake(struct pt_regs *ctx)
+{
+	u64 mntns_id;
+	u32 tid;
+	u64 ts;
+
+	mntns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	tid = (u32)bpf_get_current_pid_tgid();
+	ts = bpf_ktime_get_boot_ns();
+
+	bpf_map_update_elem(&ssl_starttime, &tid, &ts, BPF_ANY);
+	return 0;
+}
+
+SEC("uretprobe/libssl:SSL_do_handshake")
+int trace_uretprobe_libssl_SSL_do_handshake(struct pt_regs *ctx)
+{
+	struct event *event;
+	u64 mntns_id;
+	u64 pid_tgid;
+	u32 tid;
+	u64 ts;
+	u64 *tsp;
+	int ret;
+
+	mntns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	ts = bpf_ktime_get_boot_ns();
+	pid_tgid = bpf_get_current_pid_tgid();
+	tid = (u32)pid_tgid;
+
+	tsp = bpf_map_lookup_elem(&ssl_starttime, &tid);
+	if (tsp == 0)
+		return 0;
+	bpf_map_delete_elem(&ssl_starttime, &tid);
+
+	ret = PT_REGS_RC(ctx);
+	if (ret <= 0) // handshake failed
+		return 0;
+
+	event = gadget_reserve_buf(&events, BASE_EVENT_SIZE);
+	if (!event)
+		return 0;
+
+	event->mntns_id = mntns_id;
+	event->operation = libssl_SSL_DO_HANDSHAKE;
+	event->timestamp = ts;
+	event->latency_ns = ts - *tsp;
+	event->pid = pid_tgid >> 32;
+	event->tid = tid;
+	event->uid = bpf_get_current_uid_gid();
+	event->len = ret;
```

Comment:
This seems wrong, len should be left to 0 in the context of SSL_do_handshake.

Because of L258, ret is 1, see https://www.openssl.org/docs/manmaster/man3/SSL_do_handshake.html